add boringssl support

build with boringssl error:

```sh
src/alloc.c:67:5: error: call to undeclared function 'CRYPTO_set_mem_functions'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
   67 |     CRYPTO_set_mem_functions(crypto_malloc, crypto_realloc, crypto_free);
      |     ^
1 error generated.
make[2]: *** [CMakeFiles/tlsuv.dir/src/alloc.c.o] Error 1
make[2]: *** Waiting for unfinished jobs....
src/openssl/keys.c:24:10: fatal error: 'openssl/param_build.h' file not found
   24 | #include <openssl/param_build.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~
src/openssl/engine.c:32:10: fatal error: 'openssl/types.h' file not found
   32 | #include <openssl/types.h>
      |          ^~~~~~~~~~~~~~~~~

```

--- a/.github/workflows/cmake.yml
+++ b/.github/workflows/cmake.yml
@@ -37,6 +37,10 @@ jobs:
             tls: mbedtls
             target: x64-linux
             http: ON
+          - image: ubuntu
+            tls: boringssl
+            target: x64-linux
+            http: ON
           - image: macOS
             tls: openssl
             target: arm64-osx

--- a/.github/workflows/cmake.yml
+++ b/.github/workflows/cmake.yml
@@ -83,7 +83,7 @@ jobs:
       if: matrix.image == 'ubuntu'
       uses: awalsh128/cache-apt-pkgs-action@latest
       with:
-        packages: valgrind softhsm2 libc6-dbg
+        packages: valgrind softhsm2 libc6-dbg nasm
 
     - name: Install tools
       if: matrix.image == 'macOS'

--- a/.github/workflows/cmake.yml
+++ b/.github/workflows/cmake.yml
@@ -127,7 +127,7 @@ jobs:
       name: Configure build
       with:
         configurePreset: ${{ matrix.target }}
-        configurePresetAdditionalArgs: "[ `-DTLSUV_TLSLIB=${{ matrix.tls }}`, `-DTLSUV_HTTP=${{ matrix.http }}`, `-DVCPKG_MANIFEST_FEATURES='test;samples;${{ matrix.tls }}'` ]"
+        configurePresetAdditionalArgs: "[ `-DTLSUV_TLSLIB=${{ matrix.tls }}`, `-DTLSUV_HTTP=${{ matrix.http }}`, `-DVCPKG_MANIFEST_FEATURES='http,test;samples;${{ matrix.tls }}'`, `-DVCPKG_MANIFEST_NO_DEFAULT_FEATURES=ON` ]"
 
     - name: Build
       id: build

--- a/.github/workflows/cmake.yml
+++ b/.github/workflows/cmake.yml
@@ -127,7 +127,7 @@ jobs:
       name: Configure build
       with:
         configurePreset: ${{ matrix.target }}
-        configurePresetAdditionalArgs: "[ `-DTLSUV_TLSLIB=${{ matrix.tls }}`, `-DTLSUV_HTTP=${{ matrix.http }}`, `-DVCPKG_MANIFEST_FEATURES='http,test;samples;${{ matrix.tls }}'`, `-DVCPKG_MANIFEST_NO_DEFAULT_FEATURES=ON` ]"
+        configurePresetAdditionalArgs: "[ `-DTLSUV_TLSLIB=${{ matrix.tls }}`, `-DTLSUV_HTTP=${{ matrix.http }}`, `-DVCPKG_MANIFEST_FEATURES='http;test;samples;${{ matrix.tls }}'`, `-DVCPKG_MANIFEST_NO_DEFAULT_FEATURES=ON` ]"
 
     - name: Build
       id: build

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -35,7 +35,7 @@ else ()
     set(default_crypto "openssl")
 endif ()
 
-set(supported_tls_libs openssl mbedtls win32crypto)
+set(supported_tls_libs openssl mbedtls win32crypto boringssl)
 set(TLSUV_TLSLIB ${default_crypto} CACHE STRING "TLS implementation library ${supported_tls_libs}")
 
 if(NOT TLSUV_TLSLIB IN_LIST supported_tls_libs)

--- a/src/boringssl/CMakeLists.txt
+++ b/src/boringssl/CMakeLists.txt
@@ -1,0 +1,24 @@
+add_library(boringssl-impl OBJECT
+        engine.c
+        keys.c
+        keys.h
+)
+target_include_directories(boringssl-impl
+        PRIVATE ${PROJECT_SOURCE_DIR}/include)
+
+if (NOT TARGET OpenSSL::SSL)
+    find_package(OpenSSL)
+endif ()
+
+if (TARGET OpenSSL::SSL)
+    target_link_libraries(boringssl-impl PRIVATE OpenSSL::SSL)
+endif ()
+
+# BoringSSL's libssl is implemented in C++ (libcrypto stays C), so anything
+# that statically links it in - including this C-only project - needs the
+# C++ runtime on the final link line.
+if (APPLE)
+    target_link_libraries(boringssl-impl PUBLIC "c++")
+elseif (UNIX)
+    target_link_libraries(boringssl-impl PUBLIC "stdc++")
+endif ()

--- a/src/boringssl/engine.c
+++ b/src/boringssl/engine.c
@@ -715,7 +715,7 @@ static int tls_set_own_cert(tls_context* ctx, tlsuv_private_key_t key,
                 return -1;
             }
             store = crt->cert;
-            free(crt);
+            tlsuv__free(crt);
         }
     } else {
         // owned by the caller

--- a/src/boringssl/engine.c
+++ b/src/boringssl/engine.c
@@ -90,7 +90,7 @@ static void tls_free(tlsuv_engine_t self);
 static void tls_free_ctx(tls_context * ctx);
 
 static void tls_set_cert_verify(tls_context * ctx,
-                                int(*verify_f)(conststruct tlsuv_certificate_s * cert, void *v_ctx),
+                                int (*verify_f)(const struct tlsuv_certificate_s* cert, void* v_ctx),
                                 void *v_ctx);
 
 static int parse_pkcs7_certs(tlsuv_certificate_t* chain, const char* pkcs7, size_t pkcs7len);
@@ -606,7 +606,7 @@ int tls_set_partial_vfy(tls_context* ctx, int allow) {
 }
 
 static void tls_set_cert_verify(tls_context * ctx,
-                                int(*verify_f)(conststruct tlsuv_certificate_s * cert, void *v_ctx),
+                                int (*verify_f)(const struct tlsuv_certificate_s* cert, void* v_ctx),
                                 void *v_ctx) {
     struct openssl_ctx* c = (struct openssl_ctx*)ctx;
     c->cert_verify_f = verify_f;

--- a/src/boringssl/engine.c
+++ b/src/boringssl/engine.c
@@ -13,7 +13,9 @@
 // limitations under the License.
 
 #include <assert.h>
+#include <limits.h>
 #include <stdbool.h>
+#include <stdint.h>
 #include <stdlib.h>
 #include <string.h>
 #include <stdarg.h>

--- a/src/boringssl/engine.c
+++ b/src/boringssl/engine.c
@@ -799,6 +799,7 @@ static const char* tls_get_alpn(tlsuv_engine_t self) {
     unsigned int protolen;
     SSL_get0_alpn_selected(eng->ssl, &proto, &protolen);
 
+    tlsuv__free(eng->alpn);
     eng->alpn = tlsuv__calloc(1, protolen + 1);
     strncpy(eng->alpn, (const char*)proto, protolen);
     return eng->alpn;

--- a/src/boringssl/engine.c
+++ b/src/boringssl/engine.c
@@ -1,0 +1,1103 @@
+// Copyright (c) 2024. NetFoundry Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+//
+// You may obtain a copy of the License at
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdarg.h>
+#include <sys/stat.h>
+
+#include "../alloc.h"
+#include "../um_debug.h"
+
+#include <openssl/err.h>
+#include <openssl/x509.h>
+#include <openssl/ssl.h>
+#include <openssl/crypto.h>
+
+#include "keys.h"
+
+#if __APPLE__
+#include <Security/Security.h>
+#endif
+
+struct openssl_ctx {
+    tls_context api;
+    SSL_CTX* ctx;
+    int (*cert_verify_f)(const struct tlsuv_certificate_s* cert, void* v_ctx);
+    void* verify_ctx;
+    unsigned char* alpn_protocols;
+};
+
+struct openssl_engine {
+    struct tlsuv_engine_s api;
+    SSL* ssl;
+    char* alpn;
+
+    BIO* bio;
+    io_ctx io;
+    io_read read_f;
+    io_write write_f;
+
+    unsigned long error;
+};
+
+extern const char* tlsuv_get_config_path();
+
+static int is_self_signed(X509 * cert);
+static const char* name_str(const X509_NAME* n);
+static void init_ssl_context(struct openssl_ctx* c, const char* cabuf, size_t cabuf_len);
+static int tls_set_own_cert(tls_context* ctx, tlsuv_private_key_t key,
+                            tlsuv_certificate_t cert);
+
+static int set_ca_bundle(tls_context* tls, const char* ca, size_t ca_len);
+
+tlsuv_engine_t new_boringssl_engine(tls_context* ctx, const char* host);
+static void set_io(tlsuv_engine_t, io_ctx, io_read, io_write);
+static void set_io_fd(tlsuv_engine_t, tlsuv_sock_t);
+static void set_protocols(tlsuv_engine_t self, const char** protocols, int len);
+
+static tls_handshake_state tls_hs_state(tlsuv_engine_t engine);
+static tls_handshake_state
+tls_continue_hs(tlsuv_engine_t self);
+
+static const char* tls_get_alpn(tlsuv_engine_t self);
+
+static int tls_write(tlsuv_engine_t self, const char* data, size_t data_len);
+
+static int tls_read(tlsuv_engine_t self, char*, size_t*, size_t);
+
+static int tls_close(tlsuv_engine_t self);
+
+static int tls_reset(tlsuv_engine_t self);
+
+static const char* tls_lib_version();
+static const char* tls_eng_error(tlsuv_engine_t self);
+
+static void tls_free(tlsuv_engine_t self);
+static void tls_free_ctx(tls_context * ctx);
+
+static void tls_set_cert_verify(tls_context * ctx,
+                                int(*verify_f)(conststruct tlsuv_certificate_s * cert, void *v_ctx),
+                                void *v_ctx);
+
+static int parse_pkcs7_certs(tlsuv_certificate_t* chain, const char* pkcs7, size_t pkcs7len);
+
+static int load_cert(tlsuv_certificate_t* cert, const char* buf, size_t buflen);
+
+static int generate_csr(tlsuv_private_key_t key, char** pem, size_t* pemlen, ...);
+
+static void msg_cb(int write_p, int version, int content_type, const void* buf, size_t len, SSL* ssl, void* arg);
+static void info_cb(const SSL* s, int where, int ret);
+
+static int tls_set_partial_vfy(tls_context* ctx, int allow);
+
+static BIO_METHOD* BIO_s_engine(void);
+
+static tls_context openssl_context_api = {
+    .version = tls_lib_version,
+    .strerror = (const char *(*)(long))tls_error,
+    .new_engine = new_boringssl_engine,
+    .free_ctx = tls_free_ctx,
+    .set_ca_bundle = set_ca_bundle,
+    .set_own_cert = tls_set_own_cert,
+    .allow_partial_chain = tls_set_partial_vfy,
+    .set_cert_verify = tls_set_cert_verify,
+    .parse_pkcs7_certs = parse_pkcs7_certs,
+    .generate_key = gen_key,
+    .load_key = load_key,
+    // no PKCS11/keychain support in the BoringSSL backend
+    .load_cert = load_cert,
+    .generate_csr_to_pem = generate_csr,
+};
+
+
+static struct tlsuv_engine_s openssl_engine_api = {
+    .set_io = set_io,
+    .set_io_fd = set_io_fd,
+    .set_protocols = set_protocols,
+    .handshake_state = tls_hs_state,
+    .handshake = tls_continue_hs,
+    .get_alpn = tls_get_alpn,
+    .close = tls_close,
+    .write = tls_write,
+    .read = tls_read,
+    .reset = tls_reset,
+    .free = tls_free,
+    .strerror = tls_eng_error,
+};
+
+int configure_boringssl() {
+    // BoringSSL has no OSSL_LIB_CTX/provider concept, so there is no
+    // config file to (re)load here.
+    return 0;
+}
+
+static const char* tls_lib_version() {
+    return "BoringSSL";
+}
+
+const char* tls_error(unsigned long code) {
+    static char err_buf[32];
+    if (code == 0) return "no error";
+
+    const char* lib_err = ERR_lib_error_string(code);
+    if (lib_err) return lib_err;
+
+    snprintf(err_buf, sizeof(err_buf), "error[%lX]", code);
+    return err_buf;
+}
+
+static const char* tls_eng_error(tlsuv_engine_t self) {
+    struct openssl_engine* e = (struct openssl_engine*)self;
+    const char* err = ERR_reason_error_string(e->error);
+    return err;
+}
+
+tls_context* new_boringssl_ctx(const char* ca, size_t ca_len) {
+    OPENSSL_init_ssl(OPENSSL_INIT_SSL_DEFAULT, NULL);
+
+    struct openssl_ctx* c = tlsuv__calloc(1, sizeof(struct openssl_ctx));
+    c->api = openssl_context_api;
+    init_ssl_context(c, ca, ca_len);
+
+    return &c->api;
+}
+
+static X509_STORE* load_certs(const char* buf, size_t buf_len) {
+    X509_STORE* certs = X509_STORE_new();
+    X509* c;
+
+    // try as file
+    struct stat fstat;
+    if (stat(buf, &fstat) == 0) {
+        if (fstat.st_mode & S_IFREG) {
+            if (!X509_STORE_load_locations(certs, buf, NULL)) {
+                UM_LOG(ERR, "failed to load certs from [%s]", buf);
+            }
+        } else if (fstat.st_mode & S_IFDIR) {
+            // BoringSSL does not support directory-based (hashed) CA
+            // bundles the way OpenSSL does (no X509_STORE_load_path);
+            // callers must supply a file or PEM buffer instead.
+            UM_LOG(ERR, "directory-based CA bundle[%s] is not supported with the BoringSSL backend", buf);
+        } else {
+            UM_LOG(ERR, "cert bundle[%s] is not a regular file", buf);
+        }
+    } else {
+        // try as PEM
+        BIO* crt_bio = BIO_new_mem_buf(buf, (int)buf_len);
+        while ((c = PEM_read_bio_X509(crt_bio, NULL, NULL, NULL)) != NULL) {
+            int root = is_self_signed(c);
+            UM_LOG(VERB, "%s root[%s]",
+                   name_str(X509_get_subject_name(c)), root ? "true" : "false");
+            X509_STORE_add_cert(certs, c);
+            X509_free(c);
+        }
+        BIO_free(crt_bio);
+    }
+    return certs;
+}
+
+static int load_cert(tlsuv_certificate_t* cert, const char* buf, size_t buflen) {
+    X509_STORE* store = load_certs(buf, buflen);
+
+    STACK_OF(X509_OBJECT) * certs = X509_STORE_get0_objects(store);
+    int count = sk_X509_OBJECT_num(certs);
+    if (count == 0) {
+        X509_STORE_free(store);
+        return -1;
+    }
+
+    struct cert_s* crt = tlsuv__calloc(1, sizeof(*crt));
+    cert_init(crt);
+    crt->cert = store;
+    *cert = (tlsuv_certificate_t)crt;
+    return 0;
+}
+
+static int is_self_signed(X509* cert) {
+    X509_NAME* subj = X509_get_subject_name(cert);
+    X509_NAME* issuer = X509_get_issuer_name(cert);
+    if (X509_NAME_cmp(subj, issuer) != 0) {
+        return 0;
+    }
+
+    EVP_PKEY* pub = X509_get0_pubkey(cert);
+    return X509_verify(cert, pub);
+}
+
+static const char* name_str(const X509_NAME* n) {
+    static char buf[1024];
+    BIO* b = BIO_new(BIO_s_mem());
+    X509_NAME_print(b, n, 0);
+    BIO_read(b, buf, sizeof(buf));
+    BIO_free(b);
+    return buf;
+}
+
+#if __APPLE__
+static int apple_ca_verify(int pre_verify, X509_STORE_CTX* st) {
+    CFMutableArrayRef certs = CFArrayCreateMutable(kCFAllocatorDefault, 10, &kCFTypeArrayCallBacks);
+
+    STACK_OF(X509) * chain = X509_STORE_CTX_get1_chain(st);
+    for (int i = 0; i < sk_X509_num(chain); i++) {
+        X509* x = sk_X509_value(chain, i);
+        uint8_t* der = NULL;
+        int der_len = i2d_X509(x, &der);
+        CFDataRef d = CFDataCreate(kCFAllocatorDefault, der, der_len);
+        OPENSSL_free(der);
+
+        SecCertificateRef c = SecCertificateCreateWithData(kCFAllocatorDefault, d);
+        CFArrayAppendValue(certs, c);
+        CFRelease(d);
+        CFRelease(c);
+    }
+    sk_X509_pop_free(chain, X509_free);
+
+    SecTrustRef trust;
+    SecPolicyRef policy = SecPolicyCreateBasicX509();
+
+    CFErrorRef err = NULL;
+    bool result =
+        SecTrustCreateWithCertificates(certs, policy, &trust) == errSecSuccess &&
+        SecTrustEvaluateWithError(trust, &err);
+    CFRelease(trust);
+    CFRelease(policy);
+    CFRelease(certs);
+
+    return result;
+}
+#endif
+
+static int set_ca_bundle(tls_context* tls, const char* ca, size_t ca_len) {
+    struct openssl_ctx* c = (struct openssl_ctx*)tls;
+    SSL_CTX* ctx = c->ctx;
+
+    SSL_CTX_set_verify(ctx, SSL_VERIFY_PEER, NULL);
+    if (ca != NULL) {
+        X509_STORE* store = load_certs(ca, ca_len);
+        SSL_CTX_set0_verify_cert_store(ctx, store);
+    } else {
+        // try loading default CA stores
+#if __APPLE__
+        // Apple deprecated all access to system CA roots store on macOS and iOS(was never available).
+        // use native Apple trust verification
+        SSL_CTX_set_verify(ctx, SSL_VERIFY_PEER, apple_ca_verify);
+#else
+        SSL_CTX_set_default_verify_paths(ctx);
+#endif
+    }
+    return 0;
+}
+
+static void init_ssl_context(struct openssl_ctx* c, const char* cabuf, size_t cabuf_len) {
+    const SSL_METHOD* method = TLS_client_method();
+    SSL_CTX* ctx = SSL_CTX_new(method);
+    if (ctx == NULL) {
+        ERR_print_errors_fp(stderr);
+        UM_LOG(ERR, "FATAL: failed to create SSL_CTX: %s", tls_error(ERR_get_error()));
+        abort();
+    }
+    SSL_CTX_set_app_data(ctx, c);
+    c->ctx = ctx;
+
+    set_ca_bundle((tls_context*)c, cabuf, cabuf_len);
+    SSL_CTX_set_min_proto_version(ctx, TLS1_2_VERSION);
+    SSL_CTX_set_max_proto_version(ctx, TLS1_3_VERSION);
+
+    char* tls_debug = getenv("TLS_DEBUG");
+    if (tls_debug) {
+        SSL_CTX_set_msg_callback(ctx, msg_cb);
+        SSL_CTX_set_info_callback(ctx, info_cb);
+    }
+}
+
+typedef struct string_int_pair_st {
+    const char* name;
+    int retval;
+} OPT_PAIR, STRINT_PAIR;
+
+typedef struct openssl_ctx openssl_ctx;
+
+static const char* lookup(int val, const STRINT_PAIR* list, const char* def) {
+    for (; list->name; ++list)
+        if (list->retval == val)
+            return list->name;
+    return def;
+}
+
+static STRINT_PAIR handshakes[] = {
+    {", HelloRequest", SSL3_MT_HELLO_REQUEST},
+    {", ClientHello", SSL3_MT_CLIENT_HELLO},
+    {", ServerHello", SSL3_MT_SERVER_HELLO},
+    {", HelloVerifyRequest", DTLS1_MT_HELLO_VERIFY_REQUEST},
+    {", NewSessionTicket", SSL3_MT_NEWSESSION_TICKET},
+    {", EndOfEarlyData", SSL3_MT_END_OF_EARLY_DATA},
+    {", EncryptedExtensions", SSL3_MT_ENCRYPTED_EXTENSIONS},
+    {", Certificate", SSL3_MT_CERTIFICATE},
+    {", ServerKeyExchange", SSL3_MT_SERVER_KEY_EXCHANGE},
+    {", CertificateRequest", SSL3_MT_CERTIFICATE_REQUEST},
+    {", ServerHelloDone", SSL3_MT_SERVER_DONE},
+    {", CertificateVerify", SSL3_MT_CERTIFICATE_VERIFY},
+    {", ClientKeyExchange", SSL3_MT_CLIENT_KEY_EXCHANGE},
+    {", Finished", SSL3_MT_FINISHED},
+    {", CertificateStatus", SSL3_MT_CERTIFICATE_STATUS},
+    {", SupplementalData", SSL3_MT_SUPPLEMENTAL_DATA},
+    {", KeyUpdate", SSL3_MT_KEY_UPDATE},
+#ifndef OPENSSL_NO_NEXTPROTONEG
+    {", NextProto", SSL3_MT_NEXT_PROTO},
+#endif
+    {", MessageHash", SSL3_MT_MESSAGE_HASH},
+    {NULL}
+};
+
+
+static STRINT_PAIR alert_types[] = {
+    {" close_notify", 0},
+    {" end_of_early_data", 1},
+    {" unexpected_message", 10},
+    {" bad_record_mac", 20},
+    {" decryption_failed", 21},
+    {" record_overflow", 22},
+    {" decompression_failure", 30},
+    {" handshake_failure", 40},
+    {" bad_certificate", 42},
+    {" unsupported_certificate", 43},
+    {" certificate_revoked", 44},
+    {" certificate_expired", 45},
+    {" certificate_unknown", 46},
+    {" illegal_parameter", 47},
+    {" unknown_ca", 48},
+    {" access_denied", 49},
+    {" decode_error", 50},
+    {" decrypt_error", 51},
+    {" export_restriction", 60},
+    {" protocol_version", 70},
+    {" insufficient_security", 71},
+    {" internal_error", 80},
+    {" inappropriate_fallback", 86},
+    {" user_canceled", 90},
+    {" no_renegotiation", 100},
+    {" missing_extension", 109},
+    {" unsupported_extension", 110},
+    {" certificate_unobtainable", 111},
+    {" unrecognized_name", 112},
+    {" bad_certificate_status_response", 113},
+    {" bad_certificate_hash_value", 114},
+    {" unknown_psk_identity", 115},
+    {" certificate_required", 116},
+    {NULL}
+};
+
+static STRINT_PAIR ssl_versions[] = {
+    {"SSL 3.0", SSL3_VERSION},
+    {"TLS 1.0", TLS1_VERSION},
+    {"TLS 1.1", TLS1_1_VERSION},
+    {"TLS 1.2", TLS1_2_VERSION},
+    {"TLS 1.3", TLS1_3_VERSION},
+    {"DTLS 1.0", DTLS1_VERSION},
+    {NULL}
+};
+
+static void msg_cb(int write_p, int version, int content_type, const void* buf, size_t len, SSL* ssl, void* arg) {
+    const char* str_write_p = write_p ? ">>>" : "<<<";
+
+    const char *str_content_type = "", *str_details1 = "", *str_details2 = "";
+    const char* str_version = lookup(version, ssl_versions, "???");
+
+    const unsigned char* bp = buf;
+
+    if (version == SSL3_VERSION ||
+        version == TLS1_VERSION ||
+        version == TLS1_1_VERSION ||
+        version == TLS1_2_VERSION ||
+        version == TLS1_3_VERSION ||
+        version == DTLS1_VERSION) {
+        switch (content_type) {
+        case 20:
+            str_content_type = ", ChangeCipherSpec";
+            break;
+        case 21:
+            str_content_type = ", Alert";
+            str_details1 = ", ???";
+            if (len == 2) {
+                switch (bp[0]) {
+                case 1:
+                    str_details1 = ", warning";
+                    break;
+                case 2:
+                    str_details1 = ", fatal";
+                    break;
+                }
+                str_details2 = lookup((int)bp[1], alert_types, " ???");
+            }
+            break;
+        case 22:
+            str_content_type = ", Handshake";
+            str_details1 = "???";
+            if (len > 0)
+                str_details1 = lookup((int)bp[0], handshakes, "???");
+            break;
+        case 23:
+            str_content_type = ", ApplicationData";
+            break;
+        }
+    } else if (version == 0 && content_type == SSL3_RT_HEADER) {
+        str_version = "";
+        str_content_type = "TLS Header";
+    }
+
+    UM_LOG(TRACE, "%s %s%s [length %04lx]%s%s", str_write_p, str_version,
+           str_content_type, (unsigned long)len, str_details1,
+           str_details2);
+}
+
+void info_cb(const SSL* s, int where, int ret) {
+    const char* str;
+    int w = where & ~SSL_ST_MASK;
+
+    if (w & SSL_ST_CONNECT)
+        str = "SSL_connect";
+    else if (w & SSL_ST_ACCEPT)
+        str = "SSL_accept";
+    else
+        str = "undefined";
+
+    if (where & SSL_CB_LOOP) {
+        UM_LOG(TRACE, "%s:%s", str, SSL_state_string_long(s));
+    } else if (where & SSL_CB_ALERT) {
+        str = (where & SSL_CB_READ) ? "read" : "write";
+        UM_LOG(VERB, "SSL3 alert %s:%s:%s",
+               str,
+               SSL_alert_type_string_long(ret),
+               SSL_alert_desc_string_long(ret));
+    } else if (where & SSL_CB_EXIT) {
+        if (ret == 0)
+            UM_LOG(VERB, "%s:failed in %s", str, SSL_state_string_long(s));
+        else if (ret < 0)
+            UM_LOG(VERB, "%s:error in %s", str, SSL_state_string_long(s));
+    }
+}
+
+tlsuv_engine_t new_boringssl_engine(tls_context* ctx, const char* host) {
+    struct openssl_ctx* context = (openssl_ctx*)ctx;
+
+    struct openssl_engine* engine = tlsuv__calloc(1, sizeof(struct openssl_engine));
+    engine->api = openssl_engine_api;
+
+    engine->ssl = SSL_new(context->ctx);
+
+    SSL_set_tlsext_host_name(engine->ssl, host);
+
+    // unlike OpenSSL's SSL_set1_host(), BoringSSL's does not auto-detect an
+    // IP-address-literal host and match it against iPAddress SANs - it only
+    // ever checks dNSName SANs. Detect IP literals and use the IP-specific
+    // verify param instead so hostname verification works for both forms.
+    // unlike OpenSSL's SSL_set1_host(), BoringSSL's crashes on a NULL
+    // hostname instead of treating it as "no check", so guard explicitly.
+    if (host != NULL) {
+        ASN1_OCTET_STRING* host_ip = a2i_IPADDRESS(host);
+        if (host_ip != NULL) {
+            ASN1_OCTET_STRING_free(host_ip);
+            X509_VERIFY_PARAM_set1_ip_asc(SSL_get0_param(engine->ssl), host);
+        } else {
+            SSL_set1_host(engine->ssl, host);
+        }
+    }
+
+    SSL_set_connect_state(engine->ssl);
+
+    SSL_set_app_data(engine->ssl, engine);
+
+    return &engine->api;
+}
+
+static void set_io(tlsuv_engine_t self, io_ctx io, io_read rdf, io_write wrtf) {
+    struct openssl_engine* e = (struct openssl_engine*)self;
+    assert(e->bio == NULL);
+
+#ifdef SSL_OP_IGNORE_UNEXPECTED_EOF
+    // treat unexpected EOF like normal TLS shutdown (close notify)
+    SSL_set_options(e->ssl, SSL_OP_IGNORE_UNEXPECTED_EOF);
+#endif
+    e->bio = BIO_new(BIO_s_engine());
+    BIO_set_data(e->bio, e);
+    BIO_set_init(e->bio, true);
+    SSL_set_bio(e->ssl, e->bio, e->bio);
+
+    e->io = io;
+    e->read_f = rdf;
+    e->write_f = wrtf;
+}
+
+static void set_io_fd(tlsuv_engine_t self, tlsuv_sock_t fd) {
+    struct openssl_engine* e = (struct openssl_engine*)self;
+    assert(e->bio == NULL);
+
+    e->bio = BIO_new_socket((int)fd, false);
+    SSL_set_bio(e->ssl, e->bio, e->bio);
+}
+
+static void set_protocols(tlsuv_engine_t self, const char** protocols, int len) {
+    struct openssl_engine* e = (struct openssl_engine*)self;
+
+    size_t protolen = 0;
+    for (int i = 0; i < len; i++) {
+        protolen += strlen(protocols[i]) + 1;
+    }
+
+    unsigned char* alpn_protocols = tlsuv__malloc(protolen + 1);
+    unsigned char* p = alpn_protocols;
+    for (int i = 0; i < len; i++) {
+        size_t plen = strlen(protocols[i]);
+        *p++ = (unsigned char)plen;
+        strncpy((char*)p, protocols[i], plen);
+        p += plen;
+    }
+    *p = 0;
+    SSL_set_alpn_protos(e->ssl, alpn_protocols, strlen((char*)alpn_protocols));
+    tlsuv__free(alpn_protocols);
+}
+
+static int cert_verify_cb(X509_STORE_CTX* certs, void* ctx) {
+    struct openssl_ctx* c = ctx;
+
+    X509_STORE* store = X509_STORE_new();
+    X509* crt = X509_STORE_CTX_get0_cert(certs);
+    X509_STORE_add_cert(store, crt);
+
+    char n[1024];
+    X509_NAME_oneline(X509_get_subject_name(crt), n, 1024);
+    UM_LOG(VERB, "verifying %s", n);
+
+    int rc = 1;
+    struct cert_s cert;
+    cert_init(&cert);
+    cert.cert = store;
+    if (c->cert_verify_f && c->cert_verify_f((const struct tlsuv_certificate_s*)&cert, c->verify_ctx) != 0) {
+        UM_LOG(WARN, "verify failed for certificate[%s]", n);
+        rc = 0;
+    }
+    X509_STORE_free(store);
+    return rc;
+}
+
+int tls_set_partial_vfy(tls_context* ctx, int allow) {
+    struct openssl_ctx* c = (struct openssl_ctx*)ctx;
+    X509_VERIFY_PARAM* vfy = SSL_CTX_get0_param(c->ctx);
+    if (allow) {
+        X509_VERIFY_PARAM_set_flags(vfy, X509_V_FLAG_PARTIAL_CHAIN);
+    } else {
+        X509_VERIFY_PARAM_clear_flags(vfy, X509_V_FLAG_PARTIAL_CHAIN);
+    }
+    return 0;
+}
+
+static void tls_set_cert_verify(tls_context * ctx,
+                                int(*verify_f)(conststruct tlsuv_certificate_s * cert, void *v_ctx),
+                                void *v_ctx) {
+    struct openssl_ctx* c = (struct openssl_ctx*)ctx;
+    c->cert_verify_f = verify_f;
+    c->verify_ctx = v_ctx;
+    SSL_CTX_set_verify(c->ctx, SSL_VERIFY_PEER, NULL);
+    SSL_CTX_set_cert_verify_callback(c->ctx, cert_verify_cb, c);
+}
+
+
+static void tls_free_ctx(tls_context* ctx) {
+    struct openssl_ctx* c = (struct openssl_ctx*)ctx;
+    if (c->alpn_protocols) {
+        tlsuv__free(c->alpn_protocols);
+    }
+
+    SSL_CTX_free(c->ctx);
+    tlsuv__free(c);
+}
+
+static int tls_reset(tlsuv_engine_t self) {
+    struct openssl_engine* e = (struct openssl_engine*)self;
+    ERR_clear_error();
+
+    e->bio = NULL;
+
+    if (!SSL_clear(e->ssl)) {
+        int err = SSL_get_error(e->ssl, 0);
+        UM_LOG(ERR, "error resetting TSL enging: %d(%s)", err, tls_error(err));
+        return -1;
+    }
+    return 0;
+}
+
+static void tls_free(tlsuv_engine_t self) {
+    struct openssl_engine* e = (struct openssl_engine*)self;
+    SSL_free(e->ssl);
+
+    if (e->alpn) {
+        tlsuv__free(e->alpn);
+    }
+    tlsuv__free(e);
+}
+
+
+#define SSL_OP_CHECK(op, desc) do{ \
+if ((op) != 1) { \
+        uint32_t err = ERR_get_error(); \
+        UM_LOG(ERR, "failed to " desc ": %d(%s)", err, tls_error(err)); \
+        return TLS_ERR; \
+    }} while(0)
+
+
+static X509* tls_set_cert_internal(SSL_CTX* ssl, X509_STORE* store, EVP_PKEY* pkey) {
+    STACK_OF(X509_OBJECT) * certs = X509_STORE_get0_objects(store);
+    int num = sk_X509_OBJECT_num(certs);
+
+    // Find the certificate matching the private key — X509_STORE sorts
+    // by subject name hash, not insertion order, so the leaf may not be
+    // at index 0.
+    X509* leaf = NULL;
+    int leaf_idx = -1;
+    for (int i = 0; i < num; i++) {
+        X509* x509 = X509_OBJECT_get0_X509(sk_X509_OBJECT_value(certs, i));
+        if (x509 && X509_check_private_key(x509, pkey) == 1) {
+            leaf = x509;
+            leaf_idx = i;
+            break;
+        }
+    }
+
+    if (leaf == NULL) {
+        UM_LOG(ERR, "no certificate matching the private key was found");
+        return NULL;
+    }
+
+    SSL_CTX_use_certificate(ssl, leaf);
+
+    // rest of certs go to chain
+    for (int i = 0; i < num; i++) {
+        if (i == leaf_idx) continue;
+        X509* x509 = X509_OBJECT_get0_X509(sk_X509_OBJECT_value(certs, i));
+        X509_up_ref(x509);
+        SSL_CTX_add_extra_chain_cert(ssl, x509);
+    }
+    return leaf;
+}
+
+static int tls_set_own_cert(tls_context* ctx, tlsuv_private_key_t key,
+                            tlsuv_certificate_t cert) {
+    struct openssl_ctx* c = (struct openssl_ctx*)ctx;
+    SSL_CTX* ssl = c->ctx;
+
+    SSL_CTX_use_PrivateKey(ssl, NULL);
+    SSL_CTX_use_certificate(ssl, NULL);
+    SSL_CTX_clear_chain_certs(ssl);
+
+    if (key == NULL) {
+        return 0;
+    }
+
+    struct cert_s* crt = (struct cert_s*)cert;
+    X509_STORE* store = NULL;
+    if (crt == NULL) {
+        if (key->get_certificate) {
+            if (key->get_certificate(key, (tlsuv_certificate_t*)&crt) != 0) {
+                return -1;
+            }
+            store = crt->cert;
+            free(crt);
+        }
+    } else {
+        // owned by the caller
+        store = crt->cert;
+        X509_STORE_up_ref(crt->cert);
+    }
+
+    if (store == NULL) {
+        return -1;
+    }
+
+    // OpenSSL requires setting certificate before private key
+    // https://www.openssl.org/docs/man3.0/man3/SSL_CTX_use_PrivateKey.html
+    struct priv_key_s* pk = (struct priv_key_s*)key;
+    X509* leaf = tls_set_cert_internal(ssl, store, pk->pkey);
+    X509_STORE_free(store);
+
+    if (leaf == NULL) {
+        UM_LOG(ERR, "failed to find leaf certificate matching private key");
+        return TLS_ERR;
+    }
+
+    SSL_OP_CHECK(SSL_CTX_use_PrivateKey(ssl, pk->pkey), "set private key");
+    return 0;
+}
+
+
+static tls_handshake_state tls_hs_state(tlsuv_engine_t engine) {
+    struct openssl_engine* eng = (struct openssl_engine*)engine;
+    int state = SSL_get_state(eng->ssl);
+    switch (state) {
+    case TLS_ST_OK: return TLS_HS_COMPLETE;
+    case TLS_ST_BEFORE: return TLS_HS_BEFORE;
+    default: return TLS_HS_CONTINUE;
+    }
+}
+
+static int print_err_cb(const char* e, size_t len, void* v) {
+    UM_LOG(WARN, "%.*s", (int)len, e);
+    return 1;
+}
+
+static tls_handshake_state
+tls_continue_hs(tlsuv_engine_t self) {
+    struct openssl_engine* eng = (struct openssl_engine*)self;
+    ERR_clear_error();
+
+    int rc = SSL_do_handshake(eng->ssl);
+
+    if (rc != 1) {
+        eng->error = ERR_peek_error();
+        ERR_print_errors_cb(print_err_cb, NULL);
+    }
+
+    if (rc == 1) {
+        // handshake completed
+        eng->error = 0;
+        return TLS_HS_COMPLETE;
+    }
+
+    int err = SSL_get_error(eng->ssl, rc);
+
+    if (rc == 0) {
+        // handshake encountered an error and was shutdown
+        UM_LOG(ERR, "openssl: handshake was terminated: %s", tls_error(eng->error));
+        return TLS_HS_ERROR;
+    }
+
+    if (err == SSL_ERROR_WANT_READ || err == SSL_ERROR_WANT_WRITE) {
+        return TLS_HS_CONTINUE;
+    } else {
+        // something else is wrong
+        UM_LOG(ERR, "openssl: handshake was terminated: %s", tls_error(eng->error));
+        return TLS_HS_ERROR;
+    }
+}
+
+static const char* tls_get_alpn(tlsuv_engine_t self) {
+    struct openssl_engine* eng = (struct openssl_engine*)self;
+    const unsigned char* proto;
+    unsigned int protolen;
+    SSL_get0_alpn_selected(eng->ssl, &proto, &protolen);
+
+    eng->alpn = tlsuv__calloc(1, protolen + 1);
+    strncpy(eng->alpn, (const char*)proto, protolen);
+    return eng->alpn;
+}
+
+static int tls_write(tlsuv_engine_t self, const char* data, size_t data_len) {
+    struct openssl_engine* eng = (struct openssl_engine*)self;
+    ERR_clear_error();
+
+    if (data_len > INT_MAX) {
+        data_len = INT_MAX;
+    }
+
+    size_t wrote = 0;
+    while (data_len > wrote) {
+        int to_write = (int)(data_len - wrote > INT_MAX ? INT_MAX : data_len - wrote);
+        int written = SSL_write(eng->ssl, (const unsigned char*)(data + wrote), to_write);
+        if (written <= 0) {
+            int err = SSL_get_error(eng->ssl, written);
+            if (err == SSL_ERROR_WANT_WRITE) {
+                if (wrote > 0) {
+                    return (int)wrote;
+                } else {
+                    return TLS_AGAIN;
+                }
+            } else {
+                eng->error = ERR_peek_last_error();
+                UM_LOG(ERR, "openssl: write[%d] error: %s", err, tls_error(eng->error));
+                return -1;
+            }
+        }
+        wrote += (size_t)written;
+    }
+
+    return (int)wrote;
+}
+
+
+static int
+tls_read(tlsuv_engine_t self, char* out, size_t* out_bytes, size_t maxout) {
+    struct openssl_engine* eng = (struct openssl_engine*)self;
+
+    int err = SSL_ERROR_NONE;
+    uint8_t* writep = (uint8_t*)out;
+    size_t total_out = 0;
+
+    ERR_clear_error();
+    while (maxout - total_out > 0) {
+        int to_read = (int)(maxout - total_out > INT_MAX ? INT_MAX : maxout - total_out);
+        int rc = SSL_read(eng->ssl, writep, to_read);
+        if (rc <= 0) {
+            err = SSL_get_error(eng->ssl, rc);
+            break;
+        }
+
+        total_out += (size_t)rc;
+        writep += rc;
+    }
+
+    if (total_out > 0) {
+        *out_bytes = total_out;
+        return SSL_pending(eng->ssl) ? TLS_MORE_AVAILABLE : TLS_OK;
+    }
+
+    *out_bytes = 0;
+
+    if (SSL_get_shutdown(eng->ssl)) {
+        return TLS_EOF;
+    }
+
+    unsigned long err_code = ERR_peek_last_error();
+    switch (err) {
+    case SSL_ERROR_NONE:
+        // should not be here
+        UM_LOG(WARN, "openssl read: SSL_ERROR_NONE with 0 bytes read");
+        return TLS_OK;
+    case SSL_ERROR_ZERO_RETURN:
+        return TLS_EOF;
+    case SSL_ERROR_WANT_WRITE:
+    case SSL_ERROR_WANT_READ:
+        return TLS_AGAIN;
+    case SSL_ERROR_SYSCALL:
+    case SSL_ERROR_SSL:
+        eng->error = err_code;
+        UM_LOG(WARN, "openssl read[%d]: %lX/%s", err, err_code, tls_error(err_code));
+
+        return TLS_ERR;
+    default:
+        UM_LOG(WARN, "openssl read: unexpected err[%d] code[%lX]", err, err_code);
+        break;
+    }
+    return TLS_OK;
+}
+
+static int tls_close(tlsuv_engine_t self) {
+    struct openssl_engine* eng = (struct openssl_engine*)self;
+    ERR_clear_error();
+
+    int rc = SSL_shutdown(eng->ssl);
+    if (rc < 0) {
+        int err = SSL_get_error(eng->ssl, rc);
+        unsigned long err_code = ERR_peek_last_error();
+        UM_LOG(WARN, "openssl shutdown[%d]: %lX/%s", err, err_code, tls_error(err_code));
+    }
+    return 0;
+}
+
+static int parse_pkcs7_certs(tlsuv_certificate_t* chain, const char* pkcs7buf, size_t pkcs7len) {
+    // BoringSSL removed the BIO_f_base64() implementation (the declaration
+    // is kept but unimplemented), so decode the (possibly PEM-style,
+    // multi-line) base64 input with the streaming EVP_Decode* API instead.
+    EVP_ENCODE_CTX* b64_ctx = EVP_ENCODE_CTX_new();
+    EVP_DecodeInit(b64_ctx);
+
+    uint8_t* der = tlsuv__malloc(pkcs7len);
+    int total = 0, out_len = 0;
+    int rc = EVP_DecodeUpdate(b64_ctx, der, &out_len, (const uint8_t*)pkcs7buf, pkcs7len);
+    if (rc < 0) {
+        tlsuv__free(der);
+        EVP_ENCODE_CTX_free(b64_ctx);
+        return -1;
+    }
+    total += out_len;
+
+    rc = EVP_DecodeFinal(b64_ctx, der + total, &out_len);
+    EVP_ENCODE_CTX_free(b64_ctx);
+    if (rc < 0) {
+        tlsuv__free(der);
+        return -1;
+    }
+    total += out_len;
+
+    const uint8_t* p = der;
+    PKCS7* pkcs7 = d2i_PKCS7(NULL, &p, total);
+    tlsuv__free(der);
+    if (pkcs7 == NULL) {
+        return -1;
+    }
+
+    STACK_OF(X509) * certs;
+    if (PKCS7_type_is_signed(pkcs7)) {
+        certs = pkcs7->d.sign->cert;
+    } else if (PKCS7_type_is_signedAndEnveloped(pkcs7)) {
+        certs = pkcs7->d.signed_and_enveloped->cert;
+    } else {
+        PKCS7_free(pkcs7);
+        return -1;
+    }
+
+    X509_STORE* store = X509_STORE_new();
+    for (int i = 0; i < sk_X509_num(certs); i++) {
+        X509* c = sk_X509_value(certs, i);
+        X509_STORE_add_cert(store, c);
+    }
+
+    struct cert_s* c = tlsuv__calloc(1, sizeof(*c));
+    cert_init(c);
+    c->cert = store;
+    *chain = (tlsuv_certificate_t)c;
+    PKCS7_free(pkcs7);
+    return 0;
+}
+
+
+static int generate_csr(tlsuv_private_key_t key, char** pem, size_t* pemlen, ...) {
+    struct priv_key_s* privkey = (struct priv_key_s*)key;
+    int ret = 0;
+    const char* op;
+    EVP_PKEY* pk = privkey->pkey;
+    X509_REQ* req = X509_REQ_new();
+    X509_NAME* subj = X509_REQ_get_subject_name(req);
+    BIO* b = BIO_new(BIO_s_mem());
+
+
+    va_list va;
+    va_start(va, pemlen);
+    while (true) {
+        char* id = va_arg(va, char*);
+        if (id == NULL) { break; }
+
+        const uint8_t* val = va_arg(va, uint8_t *);
+        if (val == NULL) { break; }
+
+        X509_NAME_add_entry_by_txt(subj, id, MBSTRING_ASC, val, -1, -1, 0);
+    }
+    va_end(va);
+
+#define ssl_check(OP) do{ \
+op = #OP;                 \
+if((OP) == 0) {           \
+ret = ERR_get_error();    \
+goto on_error;            \
+}}while(0)
+
+    ssl_check(X509_REQ_set_pubkey(req, pk));
+    ssl_check(X509_REQ_sign(req, pk, EVP_sha256()));
+    ssl_check(PEM_write_bio_X509_REQ(b, req));
+
+on_error:
+    if (ret) {
+        UM_LOG(WARN, "%s => %s", op, tls_error(ret));
+    } else {
+        size_t len = BIO_ctrl_pending(b);
+        *pem = tlsuv__calloc(1, len + 1);
+        BIO_read(b, *pem, (int)len);
+        if (pemlen) {
+            *pemlen = len;
+        }
+    }
+
+    BIO_free(b);
+    X509_REQ_free(req);
+
+    return ret;
+}
+
+
+static int engine_bio_write(BIO* b, const char* data, int len) {
+    struct openssl_engine* e = BIO_get_data(b);
+    assert(e);
+    assert(e->write_f);
+
+    ssize_t r = e->write_f(e->io, data, (size_t)len);
+    if (r > 0) {
+        return (int)r;
+    }
+
+    if (r == TLS_AGAIN) {
+        BIO_set_retry_write(b);
+        return -1;
+    }
+
+    return (int)r;
+}
+
+static int engine_bio_read(BIO* b, char* data, int len) {
+    struct openssl_engine* e = BIO_get_data(b);
+
+    assert(e->read_f);
+
+    ssize_t rc = e->read_f(e->io, data, (size_t)len);
+    if (rc > 0) {
+        return (int)rc;
+    } else if (rc == TLS_AGAIN) {
+        BIO_set_retry_read(b);
+        return -1;
+    }
+    return (int)rc;
+}
+
+static long engine_bio_ctrl(BIO* b, int cmd, long larg, void* pargs) {
+    long ret = 0;
+
+    fflush(stderr);
+
+    switch (cmd) {
+    case BIO_CTRL_FLUSH: // 11
+    case BIO_CTRL_DGRAM_GET_PEER: // 46
+        ret = 1;
+        break;
+    case BIO_CTRL_WPENDING: // 13
+        ret = 0;
+        break;
+    case BIO_CTRL_DGRAM_QUERY_MTU: // 40
+    case BIO_CTRL_DGRAM_GET_FALLBACK_MTU: // 47
+        ret = 1500;
+        break;
+    case BIO_CTRL_PUSH: // 6
+    case BIO_CTRL_POP: // 7
+        ret = 0;
+        break;
+    default:
+        UM_LOG(WARN, "unknown cmd: BIO[%p], cmd[%d], larg[%ld]", b, cmd, larg);
+        ret = 0;
+        break;
+    }
+
+    return ret;
+}
+
+static int engine_bio_create(BIO* b) {
+    return 1;
+}
+
+static int engine_bio_destroy(BIO* b) {
+    return 1;
+}
+
+static BIO_METHOD* engine_bio_meth = NULL;
+
+BIO_METHOD* BIO_s_engine(void) {
+    if (engine_bio_meth == NULL) {
+        engine_bio_meth = BIO_meth_new(BIO_get_new_index() | BIO_TYPE_SOURCE_SINK, "BIO_s_engine");
+
+        BIO_meth_set_write(engine_bio_meth, engine_bio_write);
+        BIO_meth_set_read(engine_bio_meth, engine_bio_read);
+        BIO_meth_set_ctrl(engine_bio_meth, engine_bio_ctrl);
+        BIO_meth_set_create(engine_bio_meth, engine_bio_create);
+        BIO_meth_set_destroy(engine_bio_meth, engine_bio_destroy);
+    }
+    return engine_bio_meth;
+}

--- a/src/boringssl/keys.c
+++ b/src/boringssl/keys.c
@@ -1,0 +1,443 @@
+// Copyright (c) 2024. NetFoundry Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+//
+// You may obtain a copy of the License at
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#define OPENSSL_SUPPRESS_DEPRECATED
+
+#include <openssl/err.h>
+#include <openssl/evp.h>
+#include <openssl/pem.h>
+#include <openssl/x509.h>
+#include <openssl/x509v3.h>
+
+#include <assert.h>
+#include <stdbool.h>
+#include <time.h>
+
+#include "../um_debug.h"
+#include "keys.h"
+#include "../alloc.h"
+
+static const char* cert_to_text(const struct tlsuv_certificate_s* cert);
+static int cert_to_pem(const struct tlsuv_certificate_s* c, int full, char** pem, size_t* pemlen);
+static void cert_free(tlsuv_certificate_t c);
+static int cert_verify(const struct tlsuv_certificate_s* c, enum hash_algo md, const char* data, size_t datalen,
+                       const char* sig, size_t siglen);
+static int cert_exp(const struct tlsuv_certificate_s*, struct tm* time);
+
+static X509* find_leaf_cert(X509_STORE * store);
+
+static struct cert_s cert_API = {
+    .free = cert_free,
+    .verify = cert_verify,
+    .to_pem = cert_to_pem,
+    .get_text = cert_to_text,
+    .get_expiration = cert_exp,
+};
+
+
+static int pubkey_to_pem(tlsuv_public_key_t pub, char** pem, size_t* pemlen);
+static void pubkey_free(tlsuv_public_key_t k);
+static int pubkey_verify(tlsuv_public_key_t pk, enum hash_algo md, const char* data, size_t datalen, const char* sig,
+                         size_t siglen);
+
+static struct pub_key_s PUB_KEY_API = {
+    .free = pubkey_free,
+    .verify = pubkey_verify,
+    .to_pem = pubkey_to_pem,
+};
+
+
+static void privkey_free(tlsuv_private_key_t k);
+static tlsuv_public_key_t privkey_pubkey(tlsuv_private_key_t pk);
+static int privkey_to_pem(tlsuv_private_key_t pk, char** pem, size_t* pemlen);
+static int privkey_sign(tlsuv_private_key_t pk, enum hash_algo md,
+                        const char* data, size_t datalen, char* sig, size_t* siglen);
+
+static struct priv_key_s PRIV_KEY_API = {
+    .free = privkey_free,
+    .to_pem = privkey_to_pem,
+    .pubkey = privkey_pubkey,
+    .sign = privkey_sign,
+    // no PKCS11/keychain support in the BoringSSL backend, so no
+    // get_certificate/store_certificate hook is provided (callers
+    // already check these fields for NULL before use)
+};
+
+void pub_key_init(struct pub_key_s* pubkey) {
+    *pubkey = PUB_KEY_API;
+}
+
+static tlsuv_private_key_t new_private_key(EVP_PKEY* pkey) {
+    struct priv_key_s* private_key = tlsuv__calloc(1, sizeof(struct priv_key_s));
+    *private_key = PRIV_KEY_API;
+    private_key->pkey = pkey;
+    return (tlsuv_private_key_t)private_key;
+}
+
+void cert_init(struct cert_s* c) {
+    *c = cert_API;
+}
+
+static void pubkey_free(tlsuv_public_key_t k) {
+    struct pub_key_s* pub = (struct pub_key_s*)k;
+    EVP_PKEY_free(pub->pkey);
+    tlsuv__free(pub);
+}
+
+static int verify_ecdsa_sig(EC_KEY* ec, const EVP_MD* hash, const char* data, size_t datalen, const char* sig,
+                            size_t siglen) {
+    int rc;
+    ECDSA_SIG* ecdsa_sig = NULL;
+
+    EVP_MD_CTX* digestor = EVP_MD_CTX_new();
+    uint8_t digest[EVP_MAX_MD_SIZE];
+    unsigned int digest_len;
+    EVP_DigestInit(digestor, hash);
+    EVP_DigestUpdate(digestor, data, datalen);
+    rc = EVP_DigestFinal(digestor, digest, &digest_len);
+
+    if (rc == 1) {
+        BIGNUM* r = BN_bin2bn((const uint8_t*)sig, (int)(siglen / 2), NULL);
+        BIGNUM* s = BN_bin2bn((const uint8_t*)sig + siglen / 2, (int)siglen / 2, NULL);
+
+        ecdsa_sig = ECDSA_SIG_new();
+        ECDSA_SIG_set0(ecdsa_sig, r, s);
+        rc = ECDSA_do_verify(digest, (int)digest_len, ecdsa_sig, ec);
+    }
+
+    ECDSA_SIG_free(ecdsa_sig);
+    EVP_MD_CTX_free(digestor);
+
+    return rc == 1 ? 0 : -1;
+}
+
+int verify_signature(EVP_PKEY* pk, enum hash_algo md, const char* data, size_t datalen, const char* sig,
+                     size_t siglen) {
+    const EVP_MD* hash = NULL;
+    switch (md) {
+    case hash_SHA256: hash = EVP_sha256();
+        break;
+    case hash_SHA384: hash = EVP_sha384();
+        break;
+    case hash_SHA512: hash = EVP_sha512();
+        break;
+    default:
+        break;
+    }
+
+    EVP_MD_CTX* digestor = EVP_MD_CTX_new();
+    if (EVP_DigestVerifyInit(digestor, NULL, hash, NULL, pk) != 1 ||
+        EVP_DigestVerifyUpdate(digestor, data, datalen) != 1) {
+        UM_LOG(WARN, "failed to create digest: %s", tls_error(ERR_get_error()));
+        return -1;
+    }
+
+    int rc = EVP_DigestVerifyFinal(digestor, (const uint8_t*)sig, siglen);
+    EVP_MD_CTX_free(digestor);
+
+    if (rc != 1 && EVP_PKEY_id(pk) == EVP_PKEY_EC) {
+        const uint8_t* p = (const uint8_t*)sig;
+        ECDSA_SIG* ecdsa_sig = d2i_ECDSA_SIG(NULL, &p, (int)siglen);
+
+        // if signature is not DER encoded try verifying it as raw ECDSA signature (EC-point)
+        if (ecdsa_sig == NULL) {
+            EC_KEY* ec = EVP_PKEY_get1_EC_KEY(pk);
+            int verified = verify_ecdsa_sig(ec, hash, data, datalen, sig, siglen);
+            EC_KEY_free(ec);
+            return verified;
+        }
+
+        ECDSA_SIG_free(ecdsa_sig);
+    }
+    return (rc == 1) ? 0 : -1;
+}
+
+static int pubkey_verify(tlsuv_public_key_t pk, enum hash_algo md, const char* data, size_t datalen, const char* sig,
+                         size_t siglen) {
+    struct pub_key_s* pub = (struct pub_key_s*)pk;
+    return verify_signature(pub->pkey, md, data, datalen, sig, siglen);
+}
+
+static void privkey_free(tlsuv_private_key_t k) {
+    struct priv_key_s* priv = (struct priv_key_s*)k;
+    EVP_PKEY_free(priv->pkey);
+    tlsuv__free(priv);
+}
+
+static int privkey_sign(tlsuv_private_key_t pk, enum hash_algo md, const char* data, size_t datalen, char* sig,
+                        size_t* siglen) {
+    struct priv_key_s* priv = (struct priv_key_s*)pk;
+    int rc = 0;
+    EVP_MD_CTX* digest = EVP_MD_CTX_new();
+    EVP_PKEY_CTX* pctx = NULL;
+
+    const EVP_MD* hash = NULL;
+    switch (md) {
+    case hash_SHA256: hash = EVP_sha256();
+        break;
+    case hash_SHA384: hash = EVP_sha384();
+        break;
+    case hash_SHA512: hash = EVP_sha512();
+        break;
+    default:
+        break;
+    }
+
+    if ((EVP_DigestSignInit(digest, &pctx, hash, NULL, priv->pkey) != 1) ||
+        (EVP_DigestSignUpdate(digest, data, datalen) != 1)) {
+        unsigned long err = ERR_get_error();
+        UM_LOG(WARN, "failed to setup digest %ld/%s", err, ERR_lib_error_string(err));
+        rc = -1;
+    }
+
+    if (EVP_DigestSignFinal(digest, (uint8_t*)sig, siglen) != 1) {
+        unsigned long err = ERR_get_error();
+        UM_LOG(WARN, "failed to sign digest %ld/%s", err, ERR_lib_error_string(err));
+        rc = -1;
+    }
+
+    EVP_MD_CTX_free(digest);
+    return rc;
+}
+
+
+static tlsuv_public_key_t privkey_pubkey(tlsuv_private_key_t pk) {
+    struct priv_key_s* priv = (struct priv_key_s*)pk;
+    struct pub_key_s* pub = tlsuv__calloc(1, sizeof(*pub));
+    pub_key_init(pub);
+
+    // there is probably a more straight-forward way,
+    // but I did not find it
+    BIO* bio = BIO_new(BIO_s_mem());
+    PEM_write_bio_PUBKEY(bio, priv->pkey);
+    pub->pkey = PEM_read_bio_PUBKEY(bio, NULL, NULL, NULL);
+
+    BIO_free_all(bio);
+    return (tlsuv_public_key_t)pub;
+}
+
+static int privkey_to_pem(tlsuv_private_key_t pk, char** pem, size_t* pemlen) {
+    BIO* b = BIO_new(BIO_s_mem());
+    struct priv_key_s* privkey = (struct priv_key_s*)pk;
+
+    *pem = NULL;
+    *pemlen = 0;
+
+    if (!PEM_write_bio_PKCS8PrivateKey(b, privkey->pkey, NULL, NULL, 0, NULL, NULL)) {
+        unsigned long err = ERR_get_error();
+        UM_LOG(WARN, "failed to generate PEM for private key: %ld/%s", err, ERR_lib_error_string(err));
+    } else {
+        size_t len = BIO_ctrl_pending(b);
+        *pem = tlsuv__calloc(1, len + 1);
+        BIO_read(b, *pem, (int)len);
+        *pemlen = len;
+    }
+    BIO_free(b);
+    return *pem != NULL ? 0 : -1;
+}
+
+
+static int pubkey_to_pem(tlsuv_public_key_t pub, char** pem, size_t* pemlen) {
+    BIO* b = BIO_new(BIO_s_mem());
+    struct pub_key_s* pubkey = (struct pub_key_s*)pub;
+
+    *pem = NULL;
+    *pemlen = 0;
+
+    if (!PEM_write_bio_PUBKEY(b, pubkey->pkey)) {
+        unsigned long err = ERR_get_error();
+        UM_LOG(WARN, "failed to generate PEM for public key: %ld/%s", err, ERR_lib_error_string(err));
+    } else {
+        size_t len = BIO_ctrl_pending(b);
+        *pem = tlsuv__calloc(1, len + 1);
+        BIO_read(b, *pem, (int)len);
+        *pemlen = len;
+    }
+    BIO_free(b);
+    return *pem != NULL ? 0 : -1;
+}
+
+int load_key(tlsuv_private_key_t* key, const char* keydata, size_t keydatalen) {
+    // try file
+    BIO* kb;
+    int rc = 0;
+    FILE* kf = fopen(keydata, "r");
+    if (kf != NULL) {
+        kb = BIO_new_fp(kf, 1);
+    } else {
+        kb = BIO_new_mem_buf(keydata, (int)keydatalen);
+    }
+
+    EVP_PKEY* pk = NULL;
+    if (!PEM_read_bio_PrivateKey(kb, &pk, NULL, NULL)) {
+        unsigned long err = ERR_get_error();
+        UM_LOG(WARN, "failed to load key: %ld/%s", err, ERR_lib_error_string(err));
+        rc = -1;
+    } else {
+        *key = new_private_key(pk);
+    }
+    BIO_free(kb);
+    return rc;
+}
+
+int gen_key(tlsuv_private_key_t* key) {
+    int rc = 0;
+
+    EVP_PKEY* pk = EVP_PKEY_new();
+    EVP_PKEY_CTX* pctx = EVP_PKEY_CTX_new_id(EVP_PKEY_EC, NULL);
+    EVP_PKEY_keygen_init(pctx);
+    EVP_PKEY_CTX_set_ec_paramgen_curve_nid(pctx, NID_X9_62_prime256v1);
+
+    if (!EVP_PKEY_keygen(pctx, &pk)) {
+        uint32_t err = ERR_get_error();
+        UM_LOG(ERR, "failed to generate key: %d(%s)", err, tls_error(err));
+        rc = -1;
+        EVP_PKEY_free(pk);
+    }
+
+    if (rc == 0) {
+        *key = new_private_key(pk);
+    }
+
+    EVP_PKEY_CTX_free(pctx);
+    return rc;
+}
+
+static const char* cert_to_text(const struct tlsuv_certificate_s* cert) {
+    struct cert_s* c = (struct cert_s*)cert;
+
+    if (c->text) return c->text;
+
+    X509* x509 = find_leaf_cert(c->cert);
+    if (x509 == NULL) {
+        UM_LOG(WARN, "no leaf cert found in store");
+        return NULL;
+    }
+
+    BIO* bio = BIO_new(BIO_s_mem());
+    X509_print_ex(bio, x509, 0,
+                  X509_FLAG_NO_HEADER | X509_FLAG_NO_SIGDUMP |
+                  X509_FLAG_NO_SIGNAME);
+
+    int len = BIO_pending(bio);
+    c->text = tlsuv__malloc(len + 1);
+    BIO_read(bio, c->text, len);
+    c->text[len] = '\0'; // ensure null-termination
+    BIO_free(bio);
+
+    return c->text;
+}
+
+static void cert_free(tlsuv_certificate_t cert) {
+    struct cert_s* c = (struct cert_s*)cert;
+    X509_STORE* s = c->cert;
+    if (s != NULL) {
+        X509_STORE_free(s);
+    }
+    tlsuv__free(c->text);
+    tlsuv__free(c);
+}
+
+static X509* find_leaf_cert(X509_STORE* store) {
+    STACK_OF(X509_OBJECT) * s = X509_STORE_get0_objects(store);
+    for (int i = 0; i < sk_X509_OBJECT_num(s); i++) {
+        X509* c = X509_OBJECT_get0_X509(sk_X509_OBJECT_value(s, i));
+        if (X509_check_ca(c) != 0) {
+            continue;
+        }
+
+        uint32_t ku = X509_get_key_usage(c);
+        uint32_t xku = X509_get_extended_key_usage(c);
+        if ((ku & KU_DIGITAL_SIGNATURE) && (xku & (XKU_SSL_CLIENT | XKU_SSL_SERVER))) {
+            return c;
+        }
+    }
+    return NULL;
+}
+
+static int cert_to_pem(const struct tlsuv_certificate_s* cert, int full_chain, char** pem, size_t* pemlen) {
+    X509_STORE* store = ((struct cert_s*)cert)->cert;
+    STACK_OF(X509_OBJECT) * s = X509_STORE_get0_objects(store);
+    if (sk_X509_OBJECT_num(s) == 0) {
+        UM_LOG(WARN, "store is empty");
+        return -1;
+    }
+
+    X509* leaf = find_leaf_cert(store);
+    BIO* pembio = BIO_new(BIO_s_mem());
+    if (!full_chain) {
+        if (leaf == NULL) {
+            leaf = X509_OBJECT_get0_X509(sk_X509_OBJECT_value(s, 0));
+        }
+        if (leaf) {
+            PEM_write_bio_X509(pembio, leaf);
+        }
+    } else {
+        // write leaf cert first
+        if (leaf) {
+            PEM_write_bio_X509(pembio, leaf);
+        }
+        for (int i = 0; i < sk_X509_OBJECT_num(s); i++) {
+            X509* c = X509_OBJECT_get0_X509(sk_X509_OBJECT_value(s, i));
+            if (c != leaf) {
+                PEM_write_bio_X509(pembio, c);
+            }
+        }
+    }
+
+    *pemlen = BIO_ctrl_pending(pembio);
+    *pem = tlsuv__calloc(1, *pemlen + 1);
+    BIO_read(pembio, *pem, (int)*pemlen);
+
+    BIO_free_all(pembio);
+    return 0;
+}
+
+static int cert_verify(const struct tlsuv_certificate_s* cert, enum hash_algo md, const char* data, size_t datalen,
+                       const char* sig, size_t siglen) {
+    X509_STORE* store = ((struct cert_s*)cert)->cert;
+    X509* c = find_leaf_cert(store);
+    if (c == NULL) {
+        UM_LOG(WARN, "no leaf cert");
+        return -1;
+    }
+    EVP_PKEY* pk = X509_get_pubkey(c);
+    if (pk == NULL) {
+        unsigned long err = ERR_peek_error();
+        UM_LOG(WARN, "no pub key: %ld/%s", err, ERR_lib_error_string(err));
+    }
+    int rc = verify_signature(pk, md, data, datalen, sig, siglen);
+    EVP_PKEY_free(pk);
+    return rc;
+}
+
+static int cert_exp(const struct tlsuv_certificate_s* cert, struct tm* time) {
+    if (time == NULL || cert == NULL) {
+        return -1;
+    }
+
+    X509_STORE* store = ((struct cert_s*)cert)->cert;
+    X509* c = find_leaf_cert(store);
+    if (c == NULL) {
+        UM_LOG(WARN, "no leaf cert found");
+        return -1;
+    }
+    const ASN1_TIME* notAfter = X509_get0_notAfter(c);
+    int64_t posix_time;
+    if (!ASN1_TIME_to_posix(notAfter, &posix_time)) {
+        return -1;
+    }
+    time_t t = (time_t)posix_time;
+    return gmtime_r(&t, time) != NULL ? 0 : -1;
+}

--- a/src/boringssl/keys.c
+++ b/src/boringssl/keys.c
@@ -416,6 +416,7 @@ static int cert_verify(const struct tlsuv_certificate_s* cert, enum hash_algo md
     if (pk == NULL) {
         unsigned long err = ERR_peek_error();
         UM_LOG(WARN, "no pub key: %ld/%s", err, ERR_lib_error_string(err));
+        return -1;
     }
     int rc = verify_signature(pk, md, data, datalen, sig, siglen);
     EVP_PKEY_free(pk);

--- a/src/boringssl/keys.h
+++ b/src/boringssl/keys.h
@@ -1,0 +1,49 @@
+
+// Copyright (c) 2024. NetFoundry Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+//
+// You may obtain a copy of the License at
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef TLSUV_BORINGSSL_KEYS_H
+#define TLSUV_BORINGSSL_KEYS_H
+
+#include <tlsuv/tls_engine.h>
+
+struct cert_s {
+    TLSUV_CERT_API
+    X509_STORE* cert;
+    char* text;
+};
+
+struct pub_key_s {
+    TLSUV_PUBKEY_API
+    EVP_PKEY* pkey;
+};
+
+struct priv_key_s {
+    TLSUV_PRIVKEY_API
+    EVP_PKEY* pkey;
+};
+
+const char* tls_error(unsigned long code);
+
+void pub_key_init(struct pub_key_s* pubkey);
+
+void cert_init(struct cert_s* c);
+
+int gen_key(tlsuv_private_key_t * key);
+int load_key(tlsuv_private_key_t* key, const char* keydata, size_t keydatalen);
+
+int verify_signature(EVP_PKEY* pk, enum hash_algo md, const char* data, size_t datalen, const char* sig, size_t siglen);
+
+
+#endif//TLSUV_BORINGSSL_KEYS_H

--- a/src/openssl/engine.c
+++ b/src/openssl/engine.c
@@ -827,7 +827,7 @@ static int tls_set_own_cert(tls_context *ctx, tlsuv_private_key_t key,
                 return -1;
             }
             store = crt->cert;
-            free(crt);
+            tlsuv__free(crt);
         }
     } else {
         // owned by the caller

--- a/tests/engine_tests.cpp
+++ b/tests/engine_tests.cpp
@@ -120,6 +120,8 @@ TEST_CASE("implementation test", "[engine]") {
     CHECK_THAT(tls->version(), Catch::Matchers::StartsWith("OpenSSL"));
 #elif defined(TEST_win32crypto)
     CHECK_THAT(tls->version(), Catch::Matchers::StartsWith("win32"));
+#elif defined(TEST_boringssl)
+    CHECK_THAT(tls->version(), Catch::Matchers::StartsWith("BoringSSL"));
 #else
     FAIL("invalid engine");
 #endif

--- a/tests/key_tests.cpp
+++ b/tests/key_tests.cpp
@@ -564,7 +564,11 @@ Z8AgrJehwuXYVyJrG5Tc1vnlSUhUrK2812JyXA7tkWj/qzc=
         const char *text = cert->get_text(cert);
         CHECK(text != nullptr);
         CHECK_THAT(text, Catch::Matchers::ContainsSubstring("Subject: CN=CafSvpHp0"));
+#if !defined(TEST_boringssl)
+        // BoringSSL's X509_print_ex omits the "ASN1 OID: <curve>" line under
+        // the EC public key info that OpenSSL/mbedTLS include.
         CHECK_THAT(text, Catch::Matchers::ContainsSubstring("ASN1 OID: prime256v1"));
+#endif
         printf("Cert text:\n%s\n", text);
     }
 

--- a/vcpkg.json
+++ b/vcpkg.json
@@ -32,6 +32,12 @@
       "description": "use mbedTLS",
       "dependencies": [ "mbedtls" ]
     },
+    "boringssl": {
+      "description": "use BoringSSL",
+      "dependencies": [
+        "boringssl"
+      ]
+    },
     "http": {
       "description": "adds HTTP and websocket support",
       "dependencies": [

